### DISCUSSION TITLE
On the Problem of Rewriting Response Body

### DISCUSSION THREAD

**Discussion by cj2a7t** [Kong/kong#11470] [Help] [answered]
I develop a plugin and want to rewrite the response body: 
ngx.arg[1] = rewrite body(This must be of type string)

So I need these to be converted into strings using cjson:
local rewrite_body = cjson. encode(rewrite table)

**results of encoding are not sequential and are always different. Is there any good way to solve this?**

**Comment by hanshuebner** [comment]
Hello @cj2a7t,

from your description, it is not quite apparent to me what your problem is.  You can replace the response body by calling [`kong.response.set_raw_body`](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.response/#kongresponseset_raw_bodybody).  Did you try that?  How does it not work?

Kind regards,
Hans

**Comment by ms2008** [comment]
> results of encoding are not sequential and are always different

Do you mean you want `cjson.encode` to keep the results serialized in a fixed order?

**Comment by cj2a7t** [comment]
@ms2008 @hanshuebner  yes, for example:

local rewrite_body = {
      k1 = 1,
      k2 = 2
}
`kong.response.set_raw_body ` need the value of string type
so i should cjson.encode(rewrite_body), but  cjson.encode  can not  keep the results serialized in a fixed order.

**Reply by cj2a7t** [reply]
kong.response.set_raw_body(body)
Sets the body of the response.

The body argument must be a string and is not processed in any way. This function can’t change the Content-Length header if one was added. If you decide to use this function, the Content-Length header should also be cleared, for example in the header_filter phase.

Phases

body_filter
Parameters

**body (string): The raw body.**

**Reply by hanshuebner** [reply]
JSON does not define an order for object elements.  If your client requires a certain order of the elements in a JSON object, it is not really JSON compatible.  Why do you need a particular order to be preserved?

**Reply by cj2a7t** [reply]
For example, if I proxy an upstream API, I need to perform some specific processing on its response body, and then return its response body in the same order, because the internal API, such as the VO of the java `controller`, must be in the same order every time. , but after my rewriting, it is in a different order every time, which does not conform to the business.

**Reply by hanshuebner** [reply, marked as answer]
Once you parse the body into Lua tables, the order of the keys is lost and there is no way to retain it.  This is compliant with JSON.  If your internal API does not conform to JSON, then you might need to work with the request bodies as if they were strings and perform your rewriting by string replacement, or use a streaming JSON parser that does not turn the JSON document into Lua tables in the first place.  I am not aware of any streaming JSON parsing libraries for Lua that would work out of the box, though.  Working with strings may be easy or difficult, depending on the complexity of your JSON documents.  For specific use cases, it may be the right solution.

**Reply by cj2a7t** [reply]
There is no `LinkedHashMap` in lua, can I only implement it by myself? Are there other similar open source libraries?

**Reply by hanshuebner** [reply]
I'm not aware of any Lua based JSON library that directly does what you ask for, but maybe you can extend [Jeffrey Friedl's JSON library](http://regex.info/blog/lua/json) to preserve object key order.  Alternatively, you could use [pdjson](https://github.com/skeeto/pdjson), which is a streaming JSON parser written in C, which you can invoke through the [LuaJIT ffi](https://luajit.org/ext_ffi.html).  Working with that will probably be more difficult, though, as you'll have to do the necessary transformations while you read the JSON body, keeping track of where you are in the document.

What you're doing is not really JSON processing, so you'll probably need to implement something to fit your requirements yourself.

**Reply by ms2008** [reply]
There are some alternatives mentioned [here](https://github.com/openresty/lua-cjson/issues/66) that you might want to try yourself (you'll have to compile the cjson yourself).

**Reply by cj2a7t** [reply]
ok thanks
